pandas/src/hashtable_class_helper.pxi.in

```diff
@@ -864,3 +836,109 @@ cdef class PyObjectHashTable(HashTable):
                 count += 1
 
         return np.asarray(labels)
+
+
+cdef class MultiIndexHashTable(HashTable):
+
+    def __init__(self, size_hint=1):
+        self.table = kh_init_uint64()
+        self.mi = None
+        kh_resize_uint64(self.table, size_hint)
+
+    def __dealloc__(self):
+        if self.table is not NULL:
+            kh_destroy_uint64(self.table)
+            self.table = NULL
+
+    def __len__(self):
+        return self.table.size
+
+    def sizeof(self, deep=False):
+        """ return the size of my table in bytes """
+        return self.table.n_buckets * (sizeof(uint64_t) + # keys
+                                       sizeof(size_t) + # vals
+                                       sizeof(uint32_t)) # flags
+
+    def __contains__(self, object key):
+        cdef:
+            khiter_t k
+            uint64_t value
+
+        from pandas.tools.hashing import hash_tuples
+        key = self.mi._as_valid_indexing_key(key)
+        value = hash_tuples(key)
+        k = kh_get_uint64(self.table, key)
+        return k != self.table.n_buckets
+
+    cpdef get_item(self, object key):
+        cdef:
+            khiter_t k
+            uint64_t value
+
+        from pandas.tools.hashing import hash_tuples
+        key = self.mi._as_valid_indexing_key(key)
+        value = hash_tuples(key)
+
+        # should check nlevels == len tuple
+        k = kh_get_uint64(self.table, value)
```

Comment:
here are some references on the hashing schemes:

[python using siphash](https://www.python.org/dev/peps/pep-0456/)
[siphash](https://en.m.wikipedia.org/wiki/SipHash)
[blog by ameuer about hashing](http://www.asmeurer.com/blog/posts/what-happens-when-you-mess-with-hashing-in-python/)

So the short answer is, we don't deal with collisions at all. This is different than ``hash()`` which must be-definition as it has some idiosyncratic hashing built in (e.g. -1 is not possible), so it *must* deal with collisions.

For a unique MI, I don't think in practice this is actually a problem, and if it shows up it is a bug. A tuple *should* be hashed 1-1 to its tuples. I suppose there is a theoretical possibility that this is violated.

So I suppose I could check, and I think it would be efficient (e.g. only constructing the wanted tuples as needed), and raise if this is the case. Let me see about that.

OTOH, a duplicate MI by-definition will not work with the scheme (in fact we don't use this indexer directly for duplicated MI).

@mikegraham any light?